File: zig-version/src/scanner.zig
const std = @import("std");
const CompilerError = @import("error-reporter.zig").CompilerError;

const writer = std.io.getStdOut().writer();

pub const Token = struct {
    token_type: TokenType,
    lexeme: []const u8,
    literal: Literal,
    position: usize,

    pub fn print(self: Token) !void {
        try writer.print("{s} {s} {}\n", .{ @tagName(self.token_type), self.lexeme, self.literal });
    }
};

pub const TokenType = union(enum) {
    // Single-character tokens
    LEFT_PAREN,
    RIGHT_PAREN,
    LEFT_BRACE,
    RIGHT_BRACE,
    COMMA,
    DOT,
    MINUS,
    PLUS,
    SEMICOLON,
    SLASH,
    STAR,

    // One or two character tokens
    BANG,
    BANG_EQUAL,
    EQUAL,
    EQUAL_EQUAL,
    GREATER,
    GREATER_EQUAL,
    LESS,
    LESS_EQUAL,

    // Literals
    IDENTIFIER,
    STRING,
    NUMBER,

    // Keywords
    AND,
    CONST,
    ELSE,
    FALSE,
    FN,
    FOR,
    IF,
    LET,
    OR,
    RETURN,
    TRUE,
    WHILE,

    EOF,
};

pub const Literal = union(enum) {
    none,
    number: f64,
    string: []const u8,

    pub fn format(self: Literal, comptime _: []const u8, _: std.fmt.FormatOptions, _writer: anytype) !void {
        switch (self) {
            .none => {
                _ = try _writer.write("null");
            },
            .string => |s| {
                try _writer.print("{s}", .{s});
            },
            .number => |n| {
                try _writer.print("{d}", .{n});
            },
        }
    }
};

pub const Scanner = struct {
    source: []const u8,
    tokens: std.ArrayList(Token),
    errors: std.ArrayList(CompilerError),
    start: usize,
    current: usize,
    line: usize,

    pub fn init(source: []const u8) Scanner {
        return Scanner{
            .source = source,
            .tokens = std.ArrayList(Token).init(std.heap.page_allocator),
            .errors = std.ArrayList(CompilerError).init(std.heap.page_allocator),
            .start = 0,
            .current = 0,
            .line = 1,
        };
    }

    pub fn deinit(self: *Scanner) void {
        self.tokens.deinit();
        self.errors.deinit();
    }

    fn isAtEnd(self: Scanner) bool {
        return self.current >= self.source.len;
    }

    pub fn scanTokens(self: *Scanner) !void {
        while (!self.isAtEnd()) {
            self.start = self.current;
            try self.scanToken();
        }

        try self.addToken(.EOF, null);
    }

    fn scanToken(self: *Scanner) !void {
        const c = self.advance();
        switch (c) {
            '(' => try self.addToken(.LEFT_PAREN, null),
            ')' => try self.addToken(.RIGHT_PAREN, null),
            '{' => try self.addToken(.LEFT_BRACE, null),
            '}' => try self.addToken(.RIGHT_BRACE, null),
            ',' => try self.addToken(.COMMA, null),
            '.' => try self.addToken(.DOT, null),
            '-' => try self.addToken(.MINUS, null),
            '+' => try self.addToken(.PLUS, null),
            ';' => try self.addToken(.SEMICOLON, null),
            '/' => try self.addToken(.SLASH, null),
            '*' => try self.addToken(.STAR, null),
            ' ', '\t', '\r' => {},
            '\n' => self.line += 1,
            else => {
                if (std.ascii.isDigit(c)) {
                    try self.scanNumber();
                }
                try self.addError("Unexpected character");
            },
        }
    }

    fn scanNumber(self: *Scanner) !void {
        while (std.ascii.isDigit(self.peek()) or self.peek() == '_' and std.ascii.isDigit(self.peekNext())) {
            _ = self.advance();
        }

        if (self.peek() == '.' and std.ascii.isDigit(self.peekNext())) {
            _ = self.advance();
            while (std.ascii.isDigit(self.peek()) or self.peek() == '_' and std.ascii.isDigit(self.peekNext())) {
                _ = self.advance();
            }
        }

        try self.addToken(.NUMBER, Literal{ .number = std.fmt.parseFloat(f64, self.source[self.start..self.current]) catch unreachable });
    }

    fn addToken(self: *Scanner, token_type: TokenType, literal: ?Literal) !void {
        try self.tokens.append(Token{
            .token_type = token_type,
            .lexeme = self.source[self.start..self.current],
            .literal = literal orelse .none,
            .position = self.start,
        });
    }

    fn addError(self: *Scanner, message: []const u8) !void {
        try self.errors.append(CompilerError{
            .position = self.start,
            .length = self.current - self.start,
            .message = message,
        });
    }

    fn advance(self: *Scanner) u8 {
        const c = self.source[self.current];
        self.current += 1;
        return c;
    }

    fn peek(self: Scanner) u8 {
        if (self.isAtEnd()) {
            return 0;
        }
        return self.source[self.current];
    }

    fn peekNext(self: Scanner) u8 {
        if (self.current + 1 >= self.source.len) {
            return 0;
        }
        return self.source[self.current + 1];
    }
};

File: ts-version/src/scanner.ts
import type { CompilerError, ErrorCode } from "./errors.ts";
import type { Token, TokenType } from "./tokens.ts";

const keywords: Record<string, TokenType> = {
  true: "TRUE",
  false: "FALSE",
  fn: "FN",
  let: "LET",
  const: "CONST",
  return: "RETURN",
  if: "IF",
  else: "ELSE",
  while: "WHILE",
  for: "FOR",
};

export function scanTokens(source: string): [Token[], CompilerError[]] {
  const ctx = createScannerContext(source);

  while (!isAtEnd(ctx)) {
    setStart(ctx);
    scanToken(ctx);
  }
  addToken(ctx, "EOF");

  return [ctx.tokens, ctx.errors];
}

export function scanToken(ctx: ScannerContext): void {
  const c = advance(ctx);

  switch (c) {
    case "(":
      addToken(ctx, "LEFT_PAREN");
      break;
    case ")":
      addToken(ctx, "RIGHT_PAREN");
      break;
    case "{":
      addToken(ctx, "LEFT_BRACE");
      break;
    case "}":
      addToken(ctx, "RIGHT_BRACE");
      break;
    case ",":
      addToken(ctx, "COMMA");
      break;
    case ".":
      addToken(ctx, "DOT");
      break;
    case "-":
      addToken(ctx, "MINUS");
      break;
    case "+":
      addToken(ctx, "PLUS");
      break;
    case ";":
      addToken(ctx, "SEMICOLON");
      break;
    case "/":
      if (match(ctx, "/")) {
        // A comment goes until the end of the line and it's ignored
        while (peek(ctx) !== "\n" && !isAtEnd(ctx)) {
          advance(ctx);
        }
      } else {
        addToken(ctx, "SLASH");
      }
      break;
    case "*":
      addToken(ctx, "STAR");
      break;
    case "!":
      addToken(ctx, match(ctx, "=") ? "BANG_EQUAL" : "BANG");
      break;
    case "=":
      addToken(ctx, match(ctx, "=") ? "EQUAL_EQUAL" : "EQUAL");
      break;
    case "<":
      addToken(ctx, match(ctx, "=") ? "LESS_EQUAL" : "LESS");
      break;
    case ">":
      addToken(ctx, match(ctx, "=") ? "GREATER_EQUAL" : "GREATER");
      break;
    case '"':
      scanString(ctx);
      break;
    case " ":
    case "\t":
    case "\r":
    case "\n":
      break;
    default:
      if (isDigit(c)) {
        scanNumber(ctx);
      } else if (isAlpha(c)) {
        scanIdentifier(ctx);
      } else {
        addErrorAndRecover(
          ctx,
          "UnexpectedCharacter",
          `"${c}" is not a valid character`,
        );
      }
  }
}

function scanString(ctx: ScannerContext): void {
  let unterminated = false;
  let escaped = false;
  while (escaped || peek(ctx) !== '"' && !isAtEnd(ctx)) {
    escaped = false;
    const c = advance(ctx);
    switch (c) {
      case "\\":
        escaped = true;
        break;
      case "\n":
        unterminated = true;
        return;
    }
  }

  if (isAtEnd(ctx)) {
    addErrorAndRecover(
      ctx,
      "UnterminatedString",
      "This string is missing a closing quote",
    );
    return;
  } else if (unterminated) {
    // TODO: Add a check to know if the string is multiline
    addErrorAndRecover(
      ctx,
      "UnterminatedString",
      "Multiline strings require a different syntax",
    );
  }

  advance(ctx);
  addToken(ctx, "STRING", ctx.source.slice(ctx.start + 1, ctx.current - 1));
}

const numberRegex = new RegExp(
  // NOTE: The order matters!
  [
    /0b[01]([01_]*[01])?/, // binary
    /0o[0-7]([0-7_]*[0-7])?/, // octal
    /0x[0-9a-fA-F]([0-9a-fA-F_]*[0-9a-fA-F])?/, // hex
    /[0-9]([0-9_]*[0-9])?(\.[0-9]([0-9_]*[0-9])?)?([eE][-+]?[0-9]([0-9_]*[0-9])?)?/, // integer, float and exponent
  ].map((r) => r.source).join("|"),
  "y",
);
function scanNumber(ctx: ScannerContext): void {
  numberRegex.lastIndex = ctx.start;
  const literal = numberRegex.exec(ctx.source);
  if (!literal) {
    addErrorAndRecover(
      ctx,
      "UnexpectedCharacter",
      "The format of the number is invalid",
    );
    return;
  }
  while (ctx.current < numberRegex.lastIndex) {
    advance(ctx);
  }
  addToken(ctx, "NUMBER", Number(literal[0].replaceAll("_", "")));
}

function scanIdentifier(ctx: ScannerContext): void {
  while (isAlphaNumeric(peek(ctx))) {
    advance(ctx);
  }
  const type = keywords[ctx.source.slice(ctx.start, ctx.current)];
  if (type) {
    addToken(ctx, type);
    return;
  }
  addToken(ctx, "IDENTIFIER");
}

function isAlpha(str: string): boolean {
  for (let i = 0; i < str.length; i++) {
    const code = str.charCodeAt(i);
    if (
      (code >= 65 && code <= 90) ||
      (code >= 97 && code <= 122) ||
      code === 95
    ) {
      continue;
    }
    return false;
  }
  return true;
}

function isDigit(str: string): boolean {
  for (let i = 0; i < str.length; i++) {
    const code = str.charCodeAt(i);
    if (code >= 48 && code <= 57) {
      continue;
    }
    return false;
  }
  return true;
}

function isAlphaNumeric(str: string): boolean {
  for (let i = 0; i < str.length; i++) {
    const code = str.charCodeAt(i);
    if (
      (code >= 65 && code <= 90) ||
      (code >= 97 && code <= 122) ||
      (code >= 48 && code <= 57) ||
      code === 95
    ) {
      continue;
    }
    return false;
  }
  return true;
}

// -----------------
//  Scanner Context
// -----------------

type ScannerContext = {
  source: string;
  tokens: Token[];
  errors: CompilerError[];
  start: number;
  current: number;
};

function createScannerContext(source: string): ScannerContext {
  return {
    source,
    tokens: [],
    errors: [],
    start: 0,
    current: 0,
  };
}

function addToken(
  ctx: ScannerContext,
  type: TokenType,
  literal?: number | string,
): void {
  ctx.tokens.push({
    type,
    lexeme: type === "EOF" ? "<eof>" : ctx.source.slice(ctx.start, ctx.current),
    literal,
    position: ctx.start,
  });
}

function addErrorAndRecover(
  ctx: ScannerContext,
  code: ErrorCode,
  message: string,
): void {
  ctx.errors.push({
    type: "error",
    position: ctx.start,
    length: ctx.current - ctx.start,
    code,
    message,
  });
  while (!isAtEnd(ctx) && peek(ctx) !== "\n") {
    advance(ctx);
  }
}

function isAtEnd(ctx: ScannerContext): boolean {
  return ctx.current >= ctx.source.length;
}

function setStart(ctx: ScannerContext): void {
  ctx.start = ctx.current;
}

function advance(ctx: ScannerContext): string {
  const c = ctx.source[ctx.current];
  ctx.current += 1;
  return c;
}

function match(ctx: ScannerContext, expected: string): boolean {
  if (isAtEnd(ctx)) return false;
  if (ctx.source[ctx.current] !== expected) return false;

  ctx.current += 1;
  return true;
}

function peek(ctx: ScannerContext): string {
  if (isAtEnd(ctx)) {
    return "";
  }
  return ctx.source[ctx.current];
}
